src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,316 @@
+        let filesWithChangedSetOfUnresolvedImports: Path[];
+        let filesWithInvalidatedResolutions: Map<true>;
+
+        const resolvedModuleNames = createMap<Map<ResolvedModuleWithFailedLookupLocations>>();
+        const resolvedTypeReferenceDirectives = createMap<Map<ResolvedTypeReferenceDirectiveWithFailedLookupLocations>>();
+
+        type FailedLookupLocationsWatcher = { fileWatcher: FileWatcher; refCount: number };
+        const failedLookupLocationsWatches = createMap<FailedLookupLocationsWatcher>();
+
+        return {
+            setModuleResolutionHost,
+            startRecordingFilesWithChangedResolutions,
+            finishRecordingFilesWithChangedResolutions,
+            resolveModuleNames,
+            resolveTypeReferenceDirectives,
+            invalidateResolutionOfDeletedFile,
+            invalidateResolutionOfChangedFailedLookupLocation,
+            createHasInvalidatedResolution,
+            clear
+        };
+
+        function setModuleResolutionHost(updatedHost: ModuleResolutionHost) {
+            host = updatedHost;
+        }
+
+        function clear() {
+            failedLookupLocationsWatches.forEach((failedLookupLocationWatcher, failedLookupLocationPath: Path) => {
+                log(`Watcher: FailedLookupLocations: Status: ForceClose: LocationPath: ${failedLookupLocationPath}, refCount: ${failedLookupLocationWatcher.refCount}`);
+                failedLookupLocationWatcher.fileWatcher.close();
+            });
+            failedLookupLocationsWatches.clear();
+            resolvedModuleNames.clear();
+            resolvedTypeReferenceDirectives.clear();
+        }
+
+        function startRecordingFilesWithChangedResolutions() {
+            filesWithChangedSetOfUnresolvedImports = [];
+        }
+
+        function finishRecordingFilesWithChangedResolutions() {
+            const collected = filesWithChangedSetOfUnresolvedImports;
+            filesWithChangedSetOfUnresolvedImports = undefined;
+            return collected;
+        }
+
+        function createHasInvalidatedResolution(): HasInvalidatedResolution {
+            const collected = filesWithInvalidatedResolutions;
+            filesWithInvalidatedResolutions = undefined;
+            return path => collected && collected.has(path);
+        }
+
+        function resolveModuleName(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModuleWithFailedLookupLocations {
+            const primaryResult = ts.resolveModuleName(moduleName, containingFile, compilerOptions, host);
+            // return result immediately only if it is .ts, .tsx or .d.ts
+            // otherwise try to load typings from @types
+            return (resolveWithGlobalCache && resolveWithGlobalCache(primaryResult, moduleName, compilerOptions, host)) || primaryResult;
+        }
+
+        function resolveNamesWithLocalCache<T extends NameResolutionWithFailedLookupLocations, R>(

```

Comment:
That one caches for program. This one is across the programs and something that existed in tsserrver, its just popped out from there so that -- watch can benefit from this.